### ISSUE TITLE
Singleton for Analytics Service

### ISSUE THREAD

**Issue by wangkanai** [wangkanai/Analytics#2]
The idea to ask the community regarding the core service that this project using when the app domain start and load the configuration into memory for the first time. Singleton pattern is one of well-known patterns software engineering. We right now have 6 versions of singleton patterns to consider and we would you to help us make our final selection.

### 1st not thread-safe
````
public sealed class Analytics
{
    private static Analytics instance = null;
    private Analytics()
    {
    }       
    public static Analytics Instance
    {
        get
        {
            if (instance == null)
            {
                instance = new Analytics();
            }
                return instance;
            }
     }
}
````
We know that this pattern is not thread-safe and would now work well in our multi-thread environment of serving stateless webpages. **REJECTED**

### 2nd simple thread-safety
````
public sealed class Analytics
{
	private static Analytics instance = null;
	private static readonly object padlock = new object();

	Analytics()
	{
	}

	public static Analytics Instance
	{
		get
		{
			lock (padlock)
			{
				if (instance == null)
				{
					instance = new Analytics();
				}
				return instance;
			}
		}
	}
}
````
This patterns is thread-safe by introducing a locking, but also bring in new issue of deadlock. which rise the concern in high volume http calls to the web server. **REJECTED**

### 3rd attempted thread-safety using double-check locking
````
public sealed class Analytics
{
	private static Analytics instance = null;
	private static readonly object padlock = new object();

	Analytics()
	{
	}

	public static Analytics Instance
	{
		get
		{
			if (instance == null)
			{
				lock (padlock)
				{
					if (instance == null)
					{
						instance = new Analytics();
					}
				}
			}
			return instance;
		}
	}
}
````
This pattern solved the issue of deadlock that was introduced in 2nd pattern by using double-check locking. But also it easy to get wrong and does perform any better than the pattern before it. **REJECTED** 

### 4th not quite as lazy, but thread-safe without using locks
````
public sealed class Analytics
{
	private static readonly Analytics instance = new Analytics();

	static Analytics()
	{
	}

	private Analytics()
	{
	}

	public static Analytics Instance
	{
		get
		{
			return instance;
		}
	}
}
````
This pattern is interesting because of static constructor meaning that it will execute only when an instance of the class is created or a static member is referenced, and to execute only once per AppDomain. But if you have static members other than Instance, the first reference to those members will involve creating the instance. This pattern is what we considering as **2nd SELECTION**
 
### 5th fully lazy instantiation
````
public sealed class Analytics
{
	private Analytics()
	{
	}

	public static Analytics Instance { get { return Nested.instance; } }

	private class Nested
	{
		static Nested()
		{
		}

		internal static readonly Analytics instance = new Analytics();
	}
}
````
This pattern is other one that might be interesting as due to it only instantiate when triggered by the first reference to the static member of the nested class, which only occurs in Instance. but as our current plan to develop only non-static methods in this service class. So we are planning to make this pattern as **3rd SELECTION**

### 6th not quite as lazy using volatile
````
public sealed class Analytics
{
	// Singleton
	private static volatile Analytics instance = new Analytics();
	public static Analytics Instance {get { return instance;} }		
	private Analytics() {}
	static Analytics() {}	        
}
````
This approach ensures that only one instance is created and only when the instance is needed. Built upon pattern 4 and adding the declared variable `volatile` to ensure that assignment to the instance variable completes before the instance variable can be accessed. This make its interesting as due we wanna make sure that configuration file loaded once and only once. Therefore, we are planning to make this our **1st SELECTION**

### 7th using .NET 4's Lazy<T> type
````
public sealed class Analytics
{
	private static readonly Lazy<Analytics> lazy =
		new Lazy<Analytics>(() => new Analytics());

	public static Analytics Instance { get { return lazy.Value; } }

	private Analytics()
	{
	}
}
````
.NET 4.0 introduce something really interesting is concept of `System.Lazy<T>` that make laziness really simple. But as due to we are planning to build out this library based upon .NET Core run-time and will be best wait for the library to mature it code base. **FUTURE**

We would like to hear your opinion on what pattern would best suit our project.
